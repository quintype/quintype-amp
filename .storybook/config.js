import { addDecorator } from "@storybook/react";
import { withInfo } from "@storybook/addon-info";
import { AmpDecorator } from "./decorators";

addDecorator(withInfo);
addDecorator(AmpDecorator);
