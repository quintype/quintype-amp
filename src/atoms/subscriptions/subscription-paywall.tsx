import React from "react";
import styled from "styled-components";

const StyledWrapper = styled.div`
  border-radius: 3px;
  box-shadow: 0 0 8px 0 rgba(0, 0, 0, 0.2);
  padding: 40px;
  text-align: left;
  background-color: ${(props) => props.theme.color.white};
`;
const StyledButton = styled.div`
  button {
    border-radius: 5px;
    background-color: ${(props) => props.theme.color.primaryColor};
    height: 100%;
    border: none;
    color: ${(props) => props.theme.color.white};
    padding: 10px;
    cursor: pointer;
    font: ${(props) => {
      const fontFamily = props.theme.font.family.primary;
      const fontWeight = props.theme.font.weight.bold;
      const fontSize = props.theme.font.size.xs;
      return `${fontWeight} ${fontSize} ${fontFamily}`;
    }};
  }
`;
const StyledLine = styled.p`
  button {
    border: none;
    background-color: transparent;
    font-size: ${(props) => props.theme.font.size.s};
    color: ${(props) => props.theme.color.black};
    font-weight: ${(props) => props.theme.font.weight.normal};
  }
  span {
    font-weight: ${(props) => props.theme.font.weight.bold};
    font-size: ${(props) => props.theme.font.size.s};
    color: ${(props) => props.theme.color.black};
    cursor: pointer;
  }
`;
export const SubscribeMessage = styled.span`
  color: ${(props) => props.theme.color.white};
  padding: 10px;
  font: ${(props) => {
    const fontFamily = props.theme.font.family.primary;
    const fontWeight = props.theme.font.weight.bold;
    const fontSize = props.theme.font.size.xs;
    return `${fontWeight} ${fontSize} ${fontFamily}`;
  }};
`;

export const SubscriptionPaywall = () => {
  return (
    <StyledWrapper>
      <h2>To continue reading</h2>
      <StyledButton>
        <button subscriptions-action="subscribe">
          <SubscribeMessage>Subscribe</SubscribeMessage>
        </button>
      </StyledButton>
      <StyledLine>
        <button subscriptions-action="login">
          Already a user ? <span> Log in</span>
        </button>
      </StyledLine>
    </StyledWrapper>
  );
};
