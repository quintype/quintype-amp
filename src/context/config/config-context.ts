import React from "react";
export const ConfigContext = React.createContext({});
export const ConfigProvider = ConfigContext.Provider;
export const ConfigConsumer = ConfigContext.Consumer;
