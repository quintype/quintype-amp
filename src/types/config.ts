export interface Config {
  publisherConfig: PublisherConfig;
  ampConfig: AMPConfig;
  opts?: ConfigOpts;
}

export interface PublisherConfig {
  "cdn-name": string;
  "text-direction": "ltr" | "rtl";
  "publisher-name": string;
  "publisher-id": number;
  "publisher-settings"?: object;
  webengage?: {
    // temporary. webengage should not come from here. remove ASAP
    "license-code": string;
    "website-url": string;
    "tracking-code": string;
  };
}

interface Colors {
  primary: string;
  secondary: string;
  "footer-background": string;
  "footer-text-color": string;
  "header-background": string;
  "section-text-color": string;
}

export interface AMPConfig {
  menu: {
    enabled: boolean;
  };
  "menu-groups"?: {
    default?: {
      id: number;
      slug: string;
      name: string;
      items: MenuItemTypes[];
    };
  };
  fonts: {
    primary: {
      url: string;
      family: string;
    };
    secondary: {
      url: string;
      family: string;
    };
  };
  colors: Colors;
  "logo-url": string;
  doubleclick: {
    "top-ad": DoubleClickAdTypes;
    "body-ad": DoubleClickAdTypes;
    "bottom-ad": DoubleClickAdTypes;
  };
  webengage?: {
    "license-code": string;
    "website-url": string;
    "tracking-code": string;
  };
  "comscore-tracking-vars"?: {
    c1?: number;
    c2?: number;
  };
  "gtm-id"?: string;
  "google-client-id-api": boolean;
  "invalid-elements-strategy": string;
  "google-analytics-tracking-id": string;
  "fallback-image-url": string;
}

interface DoubleClickAdTypes {
  width: number;
  height: number;
  "unit-path": string;
}

interface MenuItemTypes {
  "tag-name": string | null;
  "entity-properties": string | null;
  "collection-id": number | null;
  "entity-slug": string | null;
  "item-id": number | null;
  rank: number;
  title: string;
  "item-type": string | null;
  "section-slug": string | null;
  "tag-slug": string | null;
  id: number;
  "parent-id": number | null;
  url: string;
  "entity-name": string | null;
  "collection-slug": string | null;
  "section-name": string | null;
  data?: {
    color: string;
    link?: string;
  };
}

export interface SlotType {
  ampHtml: string;
  script?: string;
  styles?: string;
}
interface SlotsTypes {
  story?: {
    "top-slot"?: SlotType;
    "bottom-slot"?: SlotType;
  };
}

export interface ConfigOpts {
  templates?: object;
  slots?: SlotsTypes;
}
