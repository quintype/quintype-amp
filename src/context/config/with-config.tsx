import React, { useContext } from "react";
import { ConfigContext } from "./config-context";

const withConfig = (BaseComponent) => (props) => {
  const configFromContext = useContext(ConfigContext);
  return <BaseComponent config={configFromContext} {...props} />;
};

export { withConfig };
