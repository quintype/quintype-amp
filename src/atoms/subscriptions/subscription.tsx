import React from "react";
import { Helmet } from "react-helmet";
import { SubscriptionProps } from "./types";
import get from "lodash/get";

export const Subscription = ({ services, score, fallbackEntitlement, config }: SubscriptionProps) => {
  if (!get(
    config,
    ["opts", "featureConfig", "subscriptions"])) {
    return null;
  }

  return (
    <Helmet>
      <link rel="preload" as="script" href="https://cdn.ampproject.org/v0/amp-subscriptions-0.1.js" />
      <link rel="preload" as="script" href="https://cdn.ampproject.org/v0/amp-mustache-0.2.js" />
      <script type="application/json" id="amp-subscriptions">
        {JSON.stringify({
          services,
          score,
          fallbackEntitlement
        })}
      </script>
      <script
        async={undefined}
        custom-element="amp-subscriptions"
        src="https://cdn.ampproject.org/v0/amp-subscriptions-0.1.js"
      />
      <script
        async={undefined}
        custom-template="amp-mustache"
        src="https://cdn.ampproject.org/v0/amp-mustache-0.2.js"
      />
    </Helmet>
  );
};
