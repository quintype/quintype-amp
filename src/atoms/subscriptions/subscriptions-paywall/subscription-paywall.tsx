import React from "react";
import get from "lodash.get";
import { PaywallProps } from "../types";
import { Helmet } from "react-helmet";
// The user is a subscriber/non-subscriber and is logged/not logged into their account. He is Granted. Subscribe button depends whether he is  asubscriber or not. And login button too.
export const SubscriberAccessPaywall = ({ config, services, score, fallbackEntitlement }: PaywallProps) => {
  const paywallRender = get(config, ["opts", "render", "subscriptionRender", "paywallRender"], null);
  if (paywallRender) return paywallRender({ config, services, score, fallbackEntitlement });
  return (
    <>
      <Helmet>
        <style type="text/css">{`
        .StyledWrapper {
          border-radius: 3px;
          box-shadow: 0 0 8px 0 rgba(0, 0, 0, 0.2);
          padding: 40px;
          text-align: left;
          background-color: ${(props) => props.theme.color.white};
        }
        `}</style>
        <style type="text/css">{`
        .StyledText {
          font-size: ${(props) => props.theme.font.size.l};
          color: ${(props) => props.theme.color.black};
          font-weight: ${(props) => props.theme.font.weight.bold};
          padding: 26px 0 26px;
        }
        `}</style>
        <style type="text/css">{`
        .StyledContent {
          font-size: 16px;
          line-height: 1.31;
          color: #242424;
          margin-top: 0;
        }
        `}</style>
        <style type="text/css">{`
        .StyledButton {
          align-self: center;
        }
        `}</style>
        <style type="text/css">{`
        .StyledButton button {
          border-radius: 5px;
          background-color: #e84646;
          background-color: ${(props) => props.theme.color.primaryColor};
          border: none;
          color: ${(props) => props.theme.color.white};
          padding: 10px;
          margin: 10px 0 10px;
          cursor: pointer;
          font: ${(props) => {
            const fontFamily = props.theme.font.family.primary;
            const fontWeight = props.theme.font.weight.bold;
            const fontSize = props.theme.font.size.xs;
            return `${fontWeight} ${fontSize} ${fontFamily}`;
          }};
        }
        `}</style>
        <style type="text/css">{`
        .StyledLine {
          display: flex;
        }
        `}</style>
        <style type="text/css">{`
        .StyledLine  p {
          font-size: ${(props) => props.theme.font.size.xs};
          color: ${(props) => props.theme.color.black};
          font-weight: ${(props) => props.theme.font.weight.normal};
        }
        `}</style>
        <style type="text/css">{`
        .StyledLine button {
            border: none;
            background-color: transparent;
          }
           `}</style>
        <style type="text/css">{`
          .StyledLine span {
            font-size: 16px;
            font-weight: 600;
            color: #e84646;
            font-weight: ${(props) => props.theme.font.weight.bold};
            font-size: ${(props) => props.theme.font.size.s};
            color: ${(props) => props.theme.color.black};
            cursor: pointer;
          }
           `}</style>
        <style type="text/css">{`
        .SubscribeMessage {
          color: ${(props) => props.theme.color.white};
          padding: 10px;
          font-family: ${(props) => props.theme.font.family.primary};
          font-weight: ${(props) => props.theme.font.weight.bold};
          font-size: ${(props) => props.theme.font.size.xs};
        }
      `}</style>
      </Helmet>
      <div
        dangerouslySetInnerHTML={{
          __html: `<section class="StyledWrapper" subscriptions-actions subscriptions-display="NOT granted">
        <h2 class="StyledText" subscriptions-actions subscriptions-display="NOT granted">
          Get unlimited access
        </h2>
        <p class="StyledContent">In publishing and graphic design, Lorem ipsum is a placeholder text commonly used to demonstrate the visual form of a document or a typeface without relying on meaningful content</p>
        <div class="StyledButton" subscriptions-actions subscriptions-display="NOT granted">
          <button subscriptions-action="subscribe" subscriptions-display="NOT granted">
            <span class="SubscribeMessage">Subscribe</span>
          </button>
        </div>
        <div class="StyledLine" subscriptions-actions subscriptions-display="NOT granted AND NOT data.isLoggedIn">
          <p>Already a user ?</p>
          <button subscriptions-action="login" subscriptions-display="NOT granted AND NOT data.isLoggedIn">
            <span> Log in</span>
          </button>
        </div>
    </section>`
        }}
      />
    </>
  );
};
// The user has read 4 out of 5 free articles. He is granted and he is on metering.
export const MeteredPaywall = ({ config, services, score, fallbackEntitlement }: PaywallProps) => {
  const meterRender = get(config, ["opts", "render", "subscriptionRender", "meterRender"], null);
  if (meterRender) return meterRender({ config, services, score, fallbackEntitlement });
  return (
    <>
      <Helmet>
        <style type="text/css">{`
      .StyledMeter {
        display: flex;
        align-items: center;
        padding: 24px;
      }
      `}</style>
        <style type="text/css">{`
      .StyledText {
        padding: 26px;
        font-weight: 800;
        margin: 0;
      }
      `}</style>
        <style type="text/css">
          {`
      .MeteredStyledLine {
       text-align: center;
        }`}
        </style>
        <style type="text/css">{`
        .MeteredStyledLine p {
          font-size: 12px;
          font-weight: normal;
          margin: 0;
        }
        `}</style>
        <style type="text/css">{`
        .MeteredStyledLine button {
          border: none;
          background-color: transparent;
        }
        `}</style>
        <style type="text/css">{`
        .MeteredStyledLine span {
          font-weight: bold;
          font-size: 14px;
          cursor: pointer;
        }
      `}</style>
        <style type="text/css">{`
      .SubscribeMessage {
        color: #fff;
        padding: 10px;
        font-size: 14px;
        font-weight: bold;
      }
      `}</style>
        <style type="text/css">{`
        .StyledButton {
          align-self: center;
        }
      `}</style>
        <style type="text/css">
          {`
        .StyledButton button {
          border-radius: 5px;
          background-color: #f00;
          border: none;
          color: #fff;
          padding: 10px;
          margin: 10px;
          cursor: pointer;
          font-size: 14px;
          font-weight: bold;
        }`}
        </style>
      </Helmet>
      <div
        dangerouslySetInnerHTML={{
          __html: `<template class="amp-subscriptions-dialog" type="amp-mustache" subscriptions-dialog subscriptions-display="granted AND grantReason = 'METERING'">
          {{#data.numberRemaining}}
            <p class="StyledText">You are left with {{data.numberRemaining}} free articles.</p>
          {{/data.numberRemaining}}
          {{#data.isLast}}
            <div class="StyledMeter">
              <p class="StyledText">You have exceeded free stories limit for this month</p>
              <div>
                <div class="StyledButton" subscriptions-actions subscriptions-display="granted">
                  <button subscriptions-action="subscribe" subscriptions-display="granted">
                    <span class="SubscribeMessage">Subscribe</span>
                  </button>
                </div>
                {{^data.isLoggedIn}}
                <div class="MeteredStyledLine" subscriptions-actions subscriptions-display="granted">
                  <p>Already a user ?</p>
                  <button subscriptions-action="login" subscriptions-display="granted">
                    <span> Log in</span>
                  </button>
                </div>
                {{/data.isLoggedIn}}
              </div>
            </div>
            {{/data.isLast}}
      </template>`
        }}
      />
    </>
  );
};
