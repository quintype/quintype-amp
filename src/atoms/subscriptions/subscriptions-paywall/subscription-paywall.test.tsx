import React from "react";
import { shallow } from "enzyme";
import { Subscription } from "../subscription";
import { SubscriberAccessPaywall, MeteredPaywall, MeteredExhaustedPaywall } from "./subscription-paywall";
import { config } from "../../../__fixtures__/config.fixture";

describe("Subscriptions", () => {
  it("should render subscriptions", () => {
    const wrapper = shallow(<Subscription />);
    expect(wrapper).toMatchSnapshot();
  });
  it("should render Subscriber Access Paywall with Login button", () => {
    const services = [{
      authorizationUrl:
        `https://newslaundry-web.qtstage.io/api/access/v1/stories/7f3d5bdb-ec52-4047-ac0d-df4036ec974b/amp-access?key=Fxugwc1mVDyJZ2dHB58bShso&accesstype_integration_id=10&readerId=READER_ID`,
      pingbackUrl:
        `https://newslaundry-web.qtstage.io/api/access/v1/stories/7f3d5bdb-ec52-4047-ac0d-df4036ec974b/amp-access?key=Fxugwc1mVDyJZ2dHB58bShso&accesstype_integration_id=10&readerId=READER_ID`,
      actions: { login: "https://www.google.com", subscribe: "https://www.facebook.com" }
    }];
    const score = { supportsViewer: 10, isReadyToPay: 9 };
    const fallbackEntitlement = {
      source: "fallback",
      granted: false,
      grantReason: "SUBSCRIBER",
      data: { numberRemaining: 4, isLast: false, isLoggedIn: false }
    };
    const wrapper = shallow(<SubscriberAccessPaywall services={services}
      score={score}
      fallbackEntitlement={fallbackEntitlement} />);
    expect(wrapper.find("div").html()).toEqual(
      `<div><section class="StyledWrapper" subscriptions-actions subscriptions-display="NOT granted">
        <h2 class="StyledText" subscriptions-actions subscriptions-display="data.isLoggedIn">
          Get unlimited access
        </h2>
        <h2 class="StyledText" subscriptions-actions subscriptions-display="NOT data.isLoggedIn">
          Just login to continue reading
        </h2>
        <div class="StyledButton" subscriptions-actions subscriptions-display="data.isLoggedIn">
          <button subscriptions-action="subscribe" subscriptions-display="data.isLoggedIn">
            <span class="SubscribeMessage">Subscribe</span>
          </button>
        </div>
        <div class="StyledLine" subscriptions-actions subscriptions-display="NOT data.isLoggedIn">
          <p>Already a user ?</p>
          <button subscriptions-action="login" subscriptions-display="NOT data.isLoggedIn">
            <span> Log in</span>
          </button>
        </div>
    </section></div>`
    );
  });
  it("should render Subscriber Access Paywall with subscribe button", () => {
    const services = [{
      authorizationUrl:
        `https://newslaundry-web.qtstage.io/api/access/v1/stories/7f3d5bdb-ec52-4047-ac0d-df4036ec974b/amp-access?key=Fxugwc1mVDyJZ2dHB58bShso&accesstype_integration_id=10&readerId=READER_ID`,
      pingbackUrl:
        `https://newslaundry-web.qtstage.io/api/access/v1/stories/7f3d5bdb-ec52-4047-ac0d-df4036ec974b/amp-access?key=Fxugwc1mVDyJZ2dHB58bShso&accesstype_integration_id=10&readerId=READER_ID`,
      actions: { login: "https://www.google.com", subscribe: "https://www.facebook.com" }
    }];
    const score = { supportsViewer: 10, isReadyToPay: 9 };
    const fallbackEntitlement = {
      source: "fallback",
      granted: false,
      grantReason: "SUBSCRIBER",
      data: { numberRemaining: 4, isLast: false, isLoggedIn: true }
    };
    const wrapper = shallow(<SubscriberAccessPaywall services={services}
      score={score}
      fallbackEntitlement={fallbackEntitlement} />);
    expect(wrapper.find("div").html()).toEqual(
      `<div><section class="StyledWrapper" subscriptions-actions subscriptions-display="NOT granted">
        <h2 class="StyledText" subscriptions-actions subscriptions-display="data.isLoggedIn">
          Get unlimited access
        </h2>
        <h2 class="StyledText" subscriptions-actions subscriptions-display="NOT data.isLoggedIn">
          Just login to continue reading
        </h2>
        <div class="StyledButton" subscriptions-actions subscriptions-display="data.isLoggedIn">
          <button subscriptions-action="subscribe" subscriptions-display="data.isLoggedIn">
            <span class="SubscribeMessage">Subscribe</span>
          </button>
        </div>
        <div class="StyledLine" subscriptions-actions subscriptions-display="NOT data.isLoggedIn">
          <p>Already a user ?</p>
          <button subscriptions-action="login" subscriptions-display="NOT data.isLoggedIn">
            <span> Log in</span>
          </button>
        </div>
    </section></div>`
    );
  });
  it("should call paywallRender when passed", () => {
    const servicesProps = [{
      authorizationUrl:
        `https://newslaundry-web.qtstage.io/api/access/v1/stories/7f3d5bdb-ec52-4047-ac0d-df4036ec974b/amp-access?key=Fxugwc1mVDyJZ2dHB58bShso&accesstype_integration_id=10&readerId=READER_ID`,
      pingbackUrl:
        `https://newslaundry-web.qtstage.io/api/access/v1/stories/7f3d5bdb-ec52-4047-ac0d-df4036ec974b/amp-access?key=Fxugwc1mVDyJZ2dHB58bShso&accesstype_integration_id=10&readerId=READER_ID`,
      actions: { login: "https://www.google.com", subscribe: "https://www.facebook.com" }
    }];
    const scoreProps = { supportsViewer: 10, isReadyToPay: 9 };
    const fallbackEntitlementProps = {
      source: "fallback",
      granted: false,
      grantReason: "SUBSCRIBER",
      data: { numberRemaining: 4, isLast: false, isLoggedIn: false }
    };
    const paywallRender = jest.fn();
    const modifiedConfig = { ...config, opts: { render: { subscriptionRender: { paywallRender } } } };
    const wrapper = shallow(<SubscriberAccessPaywall services={servicesProps}
      score={scoreProps}
      fallbackEntitlement={fallbackEntitlementProps} config={modifiedConfig} />);
    expect(paywallRender.mock.calls.length).toBe(1);
    expect(wrapper.find("section").length).toBe(0);
  });
  it("should render Metered Paywall", () => {
    const services = [{
      authorizationUrl:
        `https://newslaundry-web.qtstage.io/api/access/v1/stories/7f3d5bdb-ec52-4047-ac0d-df4036ec974b/amp-access?key=Fxugwc1mVDyJZ2dHB58bShso&accesstype_integration_id=10&readerId=READER_ID`,
      pingbackUrl:
        `https://newslaundry-web.qtstage.io/api/access/v1/stories/7f3d5bdb-ec52-4047-ac0d-df4036ec974b/amp-access?key=Fxugwc1mVDyJZ2dHB58bShso&accesstype_integration_id=10&readerId=READER_ID`,
      actions: { login: "https://www.google.com", subscribe: "https://www.facebook.com" }
    }];
    const score = { supportsViewer: 10, isReadyToPay: 9 };
    const fallbackEntitlement = {
      source: "fallback",
      granted: true,
      grantReason: "METERING",
      data: { numberRemaining: 4, isLast: false, isLoggedIn: false }
    };
    const wrapper = shallow(<MeteredPaywall services={services}
      score={score}
      fallbackEntitlement={fallbackEntitlement} />);
    expect(wrapper.find("div").html()).toEqual(
      `<div><template class="amp-subscriptions-dialog" type="amp-mustache" subscriptions-dialog subscriptions-display="granted AND grantReason = 'METERING'">
          {{#data.numberRemaining}}
            <p class="StyledText">You are left with {{data.numberRemaining}} free articles.</p>
          {{/data.numberRemaining}}
      </template></div>`
    );
  });
  it("should call meterRender when passed", () => {
    const servicesProps = [{
      authorizationUrl:
        `https://newslaundry-web.qtstage.io/api/access/v1/stories/7f3d5bdb-ec52-4047-ac0d-df4036ec974b/amp-access?key=Fxugwc1mVDyJZ2dHB58bShso&accesstype_integration_id=10&readerId=READER_ID`,
      pingbackUrl:
        `https://newslaundry-web.qtstage.io/api/access/v1/stories/7f3d5bdb-ec52-4047-ac0d-df4036ec974b/amp-access?key=Fxugwc1mVDyJZ2dHB58bShso&accesstype_integration_id=10&readerId=READER_ID`,
      actions: { login: "https://www.google.com", subscribe: "https://www.facebook.com" }
    }];
    const scoreProps = { supportsViewer: 10, isReadyToPay: 9 };
    const fallbackEntitlementProps = {
      source: "fallback",
      granted: true,
      grantReason: "METERING",
      data: { numberRemaining: 4, isLast: false, isLoggedIn: false }
    };
    const meterRender = jest.fn();
    const modifiedConfig = {
      ...config,
      opts: { render: { subscriptionRender: { meterRender } } }
    };
    const wrapper = shallow(<MeteredPaywall services={servicesProps}
      score={scoreProps}
      fallbackEntitlement={fallbackEntitlementProps} config={modifiedConfig} />);
    expect(meterRender.mock.calls.length).toBe(1);
    expect(wrapper.find("template").length).toBe(0);
  });
  it("should render Metered Exhausted Paywall with subscribe button", () => {
    const services = [{
      authorizationUrl:
        `https://newslaundry-web.qtstage.io/api/access/v1/stories/7f3d5bdb-ec52-4047-ac0d-df4036ec974b/amp-access?key=Fxugwc1mVDyJZ2dHB58bShso&accesstype_integration_id=10&readerId=READER_ID`,
      pingbackUrl:
        `https://newslaundry-web.qtstage.io/api/access/v1/stories/7f3d5bdb-ec52-4047-ac0d-df4036ec974b/amp-access?key=Fxugwc1mVDyJZ2dHB58bShso&accesstype_integration_id=10&readerId=READER_ID`,
      actions: { login: "https://www.google.com", subscribe: "https://www.facebook.com" }
    }];
    const score = { supportsViewer: 10, isReadyToPay: 9 };
    const fallbackEntitlement = {
      source: "fallback",
      granted: true,
      grantReason: "METERING",
      data: { numberRemaining: 5, isLast: true, isLoggedIn: true }
    };
    const wrapper = shallow(<MeteredExhaustedPaywall services={services}
      score={score}
      fallbackEntitlement={fallbackEntitlement} />);
    expect(wrapper.find("div").html()).toEqual(
      `<div><template
          class="amp-subscriptions-dialog"
          type="amp-mustache"
          subscriptions-dialog
          subscriptions-display="granted AND grantReason = 'METERING'">
          <div class="StyledMeter">
          {{#data.isLast}}
            <p class="StyledText">You have exceeded free stories limit for this month</p>
          {{/data.isLast}}
            <div>
              <div class="StyledButton" subscriptions-actions subscriptions-display="granted">
                <button subscriptions-action="subscribe" subscriptions-display="granted">
                  <span class="SubscribeMessage">Subscribe</span>
                </button>
              </div>
              {{^data.isLoggedIn}}
              <div class="MeteredStyledLine" subscriptions-actions subscriptions-display="granted">
                <p>Already a user ?</p>
                <button subscriptions-action="login" subscriptions-display="granted">
                  <span> Log in</span>
                </button>
              </div>
              {{/data.isLoggedIn}}
            </div>
          </div>
          </template></div>`
    );
  });
  it("should render Metered Exhausted Paywall with subscribe and login button", () => {
    const services = [{
      authorizationUrl:
        `https://newslaundry-web.qtstage.io/api/access/v1/stories/7f3d5bdb-ec52-4047-ac0d-df4036ec974b/amp-access?key=Fxugwc1mVDyJZ2dHB58bShso&accesstype_integration_id=10&readerId=READER_ID`,
      pingbackUrl:
        `https://newslaundry-web.qtstage.io/api/access/v1/stories/7f3d5bdb-ec52-4047-ac0d-df4036ec974b/amp-access?key=Fxugwc1mVDyJZ2dHB58bShso&accesstype_integration_id=10&readerId=READER_ID`,
      actions: { login: "https://www.google.com", subscribe: "https://www.facebook.com" }
    }];
    const score = { supportsViewer: 10, isReadyToPay: 9 };
    const fallbackEntitlement = {
      source: "fallback",
      granted: true,
      grantReason: "METERING",
      data: { numberRemaining: 5, isLast: true, isLoggedIn: false }
    };
    const wrapper = shallow(<MeteredExhaustedPaywall services={services}
      score={score}
      fallbackEntitlement={fallbackEntitlement} />);
    expect(wrapper.find("div").html()).toEqual(
      `<div><template
          class="amp-subscriptions-dialog"
          type="amp-mustache"
          subscriptions-dialog
          subscriptions-display="granted AND grantReason = 'METERING'">
          <div class="StyledMeter">
          {{#data.isLast}}
            <p class="StyledText">You have exceeded free stories limit for this month</p>
          {{/data.isLast}}
            <div>
              <div class="StyledButton" subscriptions-actions subscriptions-display="granted">
                <button subscriptions-action="subscribe" subscriptions-display="granted">
                  <span class="SubscribeMessage">Subscribe</span>
                </button>
              </div>
              {{^data.isLoggedIn}}
              <div class="MeteredStyledLine" subscriptions-actions subscriptions-display="granted">
                <p>Already a user ?</p>
                <button subscriptions-action="login" subscriptions-display="granted">
                  <span> Log in</span>
                </button>
              </div>
              {{/data.isLoggedIn}}
            </div>
          </div>
          </template></div>`
    );
  });
  it("should call LastStoryMeterRender when passed", () => {
    const servicesProps = [{
      authorizationUrl:
        `https://newslaundry-web.qtstage.io/api/access/v1/stories/7f3d5bdb-ec52-4047-ac0d-df4036ec974b/amp-access?key=Fxugwc1mVDyJZ2dHB58bShso&accesstype_integration_id=10&readerId=READER_ID`,
      pingbackUrl:
        `https://newslaundry-web.qtstage.io/api/access/v1/stories/7f3d5bdb-ec52-4047-ac0d-df4036ec974b/amp-access?key=Fxugwc1mVDyJZ2dHB58bShso&accesstype_integration_id=10&readerId=READER_ID`,
      actions: { login: "https://www.google.com", subscribe: "https://www.facebook.com" }
    }];
    const scoreProps = { supportsViewer: 10, isReadyToPay: 9 };
    const fallbackEntitlementProps = {
      source: "fallback",
      granted: true,
      grantReason: "METERING",
      data: { numberRemaining: 5, isLast: true, isLoggedIn: true }
    };
    const LastStoryMeterRender = jest.fn();
    const modifiedConfig = {
      ...config,
      opts: {
        render: { subscriptionRender: { LastStoryMeterRender } }
      }
    }
      ;
    const wrapper = shallow(<MeteredExhaustedPaywall services={servicesProps}
      score={scoreProps}
      fallbackEntitlement={fallbackEntitlementProps} config={modifiedConfig} />);
    expect(LastStoryMeterRender.mock.calls.length).toBe(1);
    expect(wrapper.find("template").length).toBe(0);
  });
});
