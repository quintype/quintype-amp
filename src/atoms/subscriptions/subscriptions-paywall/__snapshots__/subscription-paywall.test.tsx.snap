// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Subscriptions should render subscriptions 1`] = `
<HelmetWrapper
  defer={true}
  encodeSpecialCharacters={true}
>
  <link
    as="script"
    href="https://cdn.ampproject.org/v0/amp-subscriptions-0.1.js"
    rel="preload"
  />
  <link
    as="script"
    href="https://cdn.ampproject.org/v0/amp-mustache-0.2.js"
    rel="preload"
  />
  <script
    id="amp-subscriptions"
    type="application/json"
  >
    {"services":[{"actions":{"login":"https://www.google.com","subscribe":"https://www.facebook.com"}}]}
  </script>
  <script
    custom-element="amp-subscriptions"
    src="https://cdn.ampproject.org/v0/amp-subscriptions-0.1.js"
  />
  <script
    custom-template="amp-mustache"
    src="https://cdn.ampproject.org/v0/amp-mustache-0.2.js"
  />
</HelmetWrapper>
`;
